cosmo_tester/framework/util.py

```diff
@@ -414,44 +402,113 @@ def generate_ca_cert(ca_cert_path, ca_key_path):
     ])
 
 
-class ExecutionWaiting(Exception):
-    """
-    raised by `wait_for_execution` if it should be retried
-    """
+class ExecutionTimeout(Exception):
+    """Raised by `wait_for_execution` if the execution takes too long."""
 
 
 class ExecutionFailed(Exception):
-    """
-    raised by `wait_for_execution` if a bad state is reached
-    """
-
-
-def retry_if_not_failed(exception):
-    return not isinstance(exception, ExecutionFailed)
+    """Raised by `wait_for_execution` if the execution fails."""
 
 
-@retrying.retry(
-    stop_max_delay=5 * 60 * 1000,
-    wait_fixed=10000,
-    retry_on_exception=retry_if_not_failed,
-)
-def wait_for_execution(manager, execution, logger):
+def wait_for_execution(manager, execution, logger, tenant=None,
+                       new_password=None, timeout=(5*60)):
     logger.info(
         'Getting workflow execution [id={execution}]'.format(
             execution=execution['id'],
         )
     )
-    execution = manager.client.executions.get(execution['id'])
-    logger.info('- execution.status = %s', execution.status)
-    if execution.status not in execution.END_STATES:
-        raise ExecutionWaiting(execution.status)
-    if execution.status != execution.TERMINATED:
-        logger.warning('Execution failed')
-        raise ExecutionFailed(
-            '{status}: {error}'.format(
-                status=execution.status,
-                error=execution['error'],
-            )
-        )
-    logger.info('Execution complete')
+    current_time = time.time()
+    # Timeout after ~5 minutes
+    timeout_time = current_time + timeout
+    password_updated = False
+    output_events(manager, execution, logger, to_time=current_time)
+
+    with set_client_tenant(manager, tenant):
+        while True:
+            try:
+                execution = manager.client.executions.get(execution['id'])
+            except UserUnauthorizedError:
+                if new_password and not password_updated:
+                    # This will happen on a restore with modified users
+                    change_rest_client_password(manager, new_password)
+                    password_updated = True
+                else:
+                    # We either shouldn't change the password, or did already.
+                    raise
+
+            prev_time = current_time
+            current_time = time.time()
+
+            output_events(manager, execution, logger, prev_time, current_time)
+
+            if time.time() >= timeout_time:
+                raise ExecutionTimeout(
+                    'Execution timed out in state: {status}'.format(
```

Comment:
execution id? maybe workflow name?

Reply:
It uses id in a previous message. I'll add it there too.